FindAllShortestPaths.cpp

```diff
@@ -0,0 +1,198 @@
+
+    distance[start] = 0;
+
+    for (int i = 0; i < Graph.size(); ++i)
+    {
+        int vertex = -1;
+
+        for (int j = 0; j < Graph.size(); ++j)
+        {
+            if (used[j] == 'f' && (vertex == -1 || distance[j] < distance[vertex]))
+            {
+                vertex = j;
+            }
+        }
+
+        if (distance[vertex] == std::numeric_limits<int>::max())
+        {
+            break;
+        }
+
+        used[vertex] = 't';
+
+        for (auto neighbor : Graph[vertex])
+        {
+            int to = neighbor.first;
+            int length = neighbor.second;
+
+            if (distance[to] >= distance[vertex] + length)
+            {
+                distance[to] = distance[vertex] + length;
+
+                DeleteNotShortestPaths(Parents, to, distance[to]);
+
+                Parents[to].push_back(std::make_pair(vertex, distance[to]));
+            }
+        }
+    }
+
+    return Parents;
+}
+
+// this function will recover all paths in `answer` using `Parents` matrix
+void RecoverPathsHelper(const std::vector<std::vector<std::pair<int, int>>>& Parents, int toVertex, int fromVertex,
+    std::vector<std::vector<int>>& answer, std::vector<int> onePath)
+{
+    // this is modified DFS algorithm which will recover all paths
+    std::stack<std::pair<int, int>> stack;
+
+    // pair< vertex in stack , index of element from `Parents[vertex]` >
+    stack.push( std::make_pair( fromVertex, 0 ) );
+
+    while (!stack.empty())
+    {
+        std::pair<int, int> s = stack.top();
+
+        onePath.push_back(s.first);
+
+        // if `s.first` is equal to toVertex then we need to save the answer and go back to see if there is more paths
+        if (s.first == toVertex)
+        {
+            std::vector<int> reversedOnePath(onePath.rbegin(), onePath.rend());
+            answer.push_back(reversedOnePath);
+        }
+
+        bool s_has_neighbors = false;
+
+        /*
+         we get from `Parents[s.first]` next index that we need. For example if `s` is pair<6, 2> then we need to push in stack Parents[6][2] and change
+         `s` to pair<6, 3>
+        */
+        if (s.second < Parents[s.first].size())
+        {
+            s_has_neighbors = true;
+            std::pair<int, int> helper = Parents[s.first][s.second];
+            helper.second = 0;
+            stack.push(helper);
+        }
+
+        if (!s_has_neighbors)
+        {
+            stack.pop();
+            if (stack.empty())
+            {
+                return;
+            }
+            s = stack.top();
+            ++s.second;
+            stack.pop();
+            stack.push(s);
+            onePath.resize(onePath.size() - 2);
+        }
+    }
+}
+
+std::vector<std::vector<int>> RecoverPaths(const std::vector<std::vector<std::pair<int, int>>>& Parents, int toVertex, int fromVertex)
+{
+    std::vector<std::vector<int>> answer;
+    std::vector<int> onePath;
```

Comment:
what  is this vector? it is unused and is passed to `RecoverPathsHelper` by value. maybe you have better to declare it in  `RecoverPathsHelper` function?

Reply:
done